opm/autodiff/StandardWell_impl.hpp

```diff
@@ -2215,4 +2220,113 @@ namespace Opm
             }
         }
     }
+
+
+
+
+    template<typename TypeTag>
+    double
+    StandardWell<TypeTag>::
+    relaxationFactorFraction(const double old_value,
+                             const double dx)
+    {
+        assert(old_value >= 0. && old_value <= 1.0);
+
+        double relaxation_factor = 1.;
+
+        // updated values without relaxation factor
+        const double possible_updated_value = old_value - dx;
+
+        // 0.95 is an experimental value remains to be optimized
+        if (possible_updated_value < 0.0) {
+            relaxation_factor = std::abs(old_value / dx) * 0.95;
+        } else if (possible_updated_value > 1.0) {
+            relaxation_factor = std::abs((1. - old_value) / dx) * 0.95;
+        }
+        // if possible_updated_value is between 0. and 1.0, then relaxation_factor
+        // remains to be one
+
+        assert(relaxation_factor >= 0. && relaxation_factor <= 1.);
+
+        return relaxation_factor;
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    double
+    StandardWell<TypeTag>::
+    determineRelaxationFactorProducer(const std::vector<double>& primary_variables,
+                                      const BVectorWell& dwells)
+    {
+        // TODO: not considering solvent yet
+        // 0.95 is a experimental value, which remains to be optimized
+        double relaxation_factor = 1.0;
```

Comment:
If you can explain to me what is the relation of the fraction of solvent to other fractions, I will try to implement it. Most likely it will not matter much.